File: CPP /Unit 05/Lec 36/Q3.cpp
Single File Programming Question
Problem Statement



Emily is an ice cream shop owner who wants to calculate the volume of different ice cream shapes to manage her inventory better. She decides to create a class hierarchy to represent various ice cream shapes, where the base class is Icecream.



This class has a pure virtual function named Quantity, which will be overridden in derived classes to compute the volume of the respective shapes. Emily uses late binding to ensure that the correct method is called at runtime based on the user’s choice of ice cream shape.



Note: 

Sphere Ice Cream Volume = (4/3) * pi * r³

Cone Ice Cream Volume = (1/3) * pi * r² * h

Use M_PI for pi and use late binding.

Input format :
The input consists of an integer choice indicating the type of ice cream shape:

for a spherical ice cream.
for a conical ice cream.
If the choice is 1, an integer r (radius) is provided.

If the choice is 2, two space-separated integers r (radius) and h (height) are provided.



Note: The input type should be an integer.

Output format :
The output should be the volume of the selected ice cream shape, rounded to two decimal points.

If the input choice is invalid, the output should be: Invalid choice



Refer to the sample output for the formatting specifications.

Code constraints :
0 ≤ r, h ≤ 1000

Sample test cases :
Input 1 :
1
4
Output 1 :
268.08
Input 2 :
2
4 5
Output 2 :
83.78
Input 3 :
4
Output 3 :
Invalid choice



#include <iostream>
#include <iomanip>
#include <cmath>
using namespace std;

class Icecream {
public:
    virtual double Quantity() = 0;
    virtual ~Icecream() {}
};

class Sphere : public Icecream {
    int r;
public:
    Sphere(int r) {
        this->r = r;
    }

    double Quantity() {
        return (4.0 / 3.0) * M_PI * r * r * r;
    }
};

class Cone : public Icecream {
    int r, h;
public:
    Cone(int r, int h) {
        this->r = r;
        this->h = h;
    }

    double Quantity() {
        return (1.0 / 3.0) * M_PI * r * r * h;
    }
};

int main() {
    int choice;
    cin >> choice;

    Icecream* icecream = nullptr;

    if (choice == 1) {
        int r;
        cin >> r;
        icecream = new Sphere(r);
    }
    else if (choice == 2) {
        int r, h;
        cin >> r >> h;
        icecream = new Cone(r, h);
    }
    else {
        cout << "Invalid choice";
        return 0;
    }

    cout << fixed << setprecision(2) << icecream->Quantity();

    delete icecream;
    return 0;
}

File: Java/Unit 03/Lec 17/Snippet/Q3.java
Single File Programming Question
Problem Statement



Shreya is a student who needs to calculate her exam percentage. Her regular percentage is calculated using the formula (obtained marks / total marks) * 100. However, as a scholarship student, she receives an additional 5% bonus on her calculated percentage. 



Write a program that uses an overridden method calculatePercentage to compute both her regular and scholarship percentages.



Regular Percentage = (obtained / total) * 100

Scholarship Percentage = Regular Percentage + 5

Input format :
The first line of input consists of an integer n, which represents the total marks in the exam.

The second line consists of an integer m, which represents the marks obtained by the student.

Output format :
The first line of output prints “Student Percentage: <percentage>%”, where the percentage is rounded to two decimal places.

The second line of output prints “Scholarship Student Percentage: <percentage>%”, where the percentage is rounded to two decimal places.



Refer to the sample output for formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

1 ≤ n ≤ 1200

1 ≤ m < n

Sample test cases :
Input 1 :
500
450
Output 1 :
Student Percentage: 90.00%
Scholarship Student Percentage: 95.00%
Input 2 :
600
300
Output 2 :
Student Percentage: 50.00%
Scholarship Student Percentage: 55.00%




import java.util.Scanner;

// Base class
class Student {

    // Method to calculate regular percentage
    double calculatePercentage(int totalMarks, int obtainedMarks) {
        return ((double) obtainedMarks / totalMarks) * 100;
    }
}

// Derived class
class ScholarshipStudent extends Student {

    // Overriding method to add scholarship bonus
    @Override
    double calculatePercentage(int totalMarks, int obtainedMarks) {

        // Calculate regular percentage
        double percentage = ((double) obtainedMarks / totalMarks) * 100;

        // Add 5% bonus
        return percentage + 5;
    }
}

// Main class
public class Main {

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        Student student = new Student();
        ScholarshipStudent scholarshipStudent = new ScholarshipStudent();

        int totalMarks = scanner.nextInt();
        int obtainedMarks = scanner.nextInt();

        double studentPercentage =
                student.calculatePercentage(totalMarks, obtainedMarks);

        double scholarshipStudentPercentage =
                scholarshipStudent.calculatePercentage(totalMarks, obtainedMarks);

        System.out.printf("Student Percentage: %.2f%%%n", studentPercentage);
        System.out.printf("Scholarship Student Percentage: %.2f%%%n", scholarshipStudentPercentage);

        scanner.close();
    }
}
